highlight/highlight
# Support for searching issues when creating them - Search and link existing issues with error or session comments.

Fixes  #4460
/claim #4460

## Summary
Introduce 'Link Issue' mode when creating issue for an error or session comment. 
All issue integrations are supported except **ClickUp**. According to [this](https://clickup.canny.io/public-api/p/using-api-get-tasks-to-search-by-title) and api documentation, they do not support searching issues via their rest api.

### UI
**On UI** this just means another radio button to determine whether the user prefers to create a new issue
or link an existing one. A search-select is provided for the user to search and choose an existing issue from their
selected issue integration.
![Screenshot 2024-02-27 at 9 42 28 AM](https://github.com/highlight/highlight/assets/119384208/293c3026-5907-47e8-9432-25547ca81a2d)
![Screenshot 2024-02-27 at 9 42 48 AM](https://github.com/highlight/highlight/assets/119384208/0bdb3c04-7d05-406c-a67f-7a09249a789c)

### DEMO
https://github.com/highlight/highlight/assets/119384208/b0116ef5-6b9d-4d97-aff0-5839986bcde5

### Backend
We needed a way to create attachment for existing issue and to search issues based on integration and a
search term/query. A sample mutation is below
```graphql
createErrorCommentForExistingIssue(
	project_id: ID!
	error_group_secure_id: String!
	text: String!
	text_for_email: String!
	tagged_admins: [SanitizedAdminInput]!
	tagged_slack_users: [SanitizedSlackChannelInput]!
	error_url: String!
	author_name: String!
	issue_url: String!
	issue_title: String!
	issue_id: String!
	integrations: [IntegrationType]!
): ErrorComment
```

Our search query also looks like this
```graphql
search_issues(
	integration_type: IntegrationType!
	project_id: ID!
	query: String!
): [IssuesSearchResult!]!
```

Introduce a toggle to allow searching and selecting of existing issues to be "attached" to
error or session comments.


## WIP

- [X] Support remaining issue integrations - Github, Height,
- [x] Discuss clickup exception - there is currently no available rest api to search clickup tasks
- [x] Support session comment issue linking
- [X] Improve Jira search results
- [X] Fix select search UI glitch
- [X] Complete self-review and e2e click-testing

## How did you test this change?

- Go to an error and start creating an issue for the error for any of the issue integrations
- Verify that you can now see to Radio options to 'Create Issue' or 'Link Issue'
- Verify that when the 'Link Issue' button is clicked the form changes to enable you to search for and select an existing issue to be linked with the error or session comment



## Are there any deployment considerations?
No. Our current issue integrations have the relevant permissions.

It might be time to consider "documenting" the required permissions for every
integration however.


## Does this work require review from our design team?
Yes. @julian-highlight checkout the screenshots above and comment.
The goal is to support

- selecting whether the user wants to link an existing issue or create a new one
- searching for and linking an existing issue.
